Mathlib/Data/Sigma/Lex.lean

```diff
@@ -0,0 +1,225 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Data.Sigma.Basic
+import Mathlib.Order.RelClasses
+
+/-!
+# Lexicographic order on a sigma type
+
+This defines the lexicographical order of two arbitrary relations on a sigma type and proves some
+lemmas about `PSigma.Lex`, which is defined in core Lean.
+
+Given a relation in the index type and a relation on each summand, the lexicographical order on the
+sigma type relates `a` and `b` if their summands are related or they are in the same summand and
+related by the summand's relation.
+
+## See also
+
+Related files are:
+* `Combinatorics.CoLex`: Colexicographic order on finite sets.
+* `Data.List.Lex`: Lexicographic order on lists.
+* `Data.Sigma.Order`: Lexicographic order on `Σ i, α i` per say.
+* `Data.PSigma.Order`: Lexicographic order on `Σ' i, α i`.
+* `Data.Prod.Lex`: Lexicographic order on `α × β`. Can be thought of as the special case of
+  `Sigma.Lex` where all summands are the same
+-/
+
+
+namespace Sigma
+
+variable {ι : Type _} {α : ι → Type _} {r r₁ r₂ : ι → ι → Prop} {s s₁ s₂ : ∀ i, α i → α i → Prop}
+  {a b : Σ i, α i}
+
+/-- The lexicographical order on a sigma type. It takes in a relation on the index type and a
+relation for each summand. `a` is related to `b` iff their summands are related or they are in the
+same summand and are related through the summand's relation. -/
+inductive Lex (r : ι → ι → Prop) (s : ∀ i, α i → α i → Prop) : ∀ _ _ : Σ i, α i, Prop
+  | left {i j : ι} (a : α i) (b : α j) : r i j → Lex r s ⟨i, a⟩ ⟨j, b⟩
+  | right {i : ι} (a b : α i) : s i a b → Lex r s ⟨i, a⟩ ⟨i, b⟩
+#align sigma.lex Sigma.Lex
+
+theorem lex_iff : Lex r s a b ↔ r a.1 b.1 ∨ ∃ h : a.1 = b.1, s b.1 (h.rec a.2) b.2 := by
+  constructor
+  · rintro (⟨a, b, hij⟩ | ⟨a, b, hab⟩)
+    · exact Or.inl hij
+
+    · exact Or.inr ⟨rfl, hab⟩
+
+
```

Comment:
I think the extra spacing here and elsewhere is not necessary. Can you remove it?